File: go/cost_aware_group_iterator.go
package turbodata

import "io"

// costAwareGroupIterator yields messages from one topic group when the reader
// runs on the cost-aware path. All bytes the group needs have been pre-fetched
// into loadedData, and the per-chunk kept-message lists have been computed at
// setup time, so this type just walks forward (or backward) through them.
type costAwareGroupIterator struct {
	isCompressed bool
	order        Order

	indexChunks   []*IndexChunk
	chunkMessages [][]messageRef // parallel to indexChunks
	chunkMsgLens  [][]int64      // parallel to indexChunks

	loadedData *LoadedBytes

	// Reused across chunks; only allocated when isCompressed.
	decompressBuf *ReusableBuffer

	// Iteration state.
	chunkIdx       int // current chunk being consumed
	msgIdx         int // current message within current chunk
	loadedChunkIdx int // which chunk's data is currently in decompressBuf (-1 = none)
}

func newCostAwareGroupIterator(
	isCompressed bool,
	order Order,
	indexChunks []*IndexChunk,
	chunkMessages [][]messageRef,
	chunkMsgLens [][]int64,
	loadedData *LoadedBytes,
) *costAwareGroupIterator {
	it := &costAwareGroupIterator{
		isCompressed:   isCompressed,
		order:          order,
		indexChunks:    indexChunks,
		chunkMessages:  chunkMessages,
		chunkMsgLens:   chunkMsgLens,
		loadedData:     loadedData,
		loadedChunkIdx: -1,
	}
	if isCompressed {
		it.decompressBuf = NewReusableBuffer()
	}
	if order == ReverseTimeOrder {
		it.chunkIdx = len(indexChunks) - 1
		if it.chunkIdx >= 0 {
			it.msgIdx = len(chunkMessages[it.chunkIdx]) - 1
		}
	} else {
		it.chunkIdx = 0
		if len(indexChunks) > 0 {
			it.msgIdx = 0
		}
	}
	return it
}

// Next returns the next (timestamp, topicId, data) record, or io.EOF when the
// group is exhausted. Returned data aliases loadedData / decompressBuf; the
// caller must copy if it needs to outlive the next call.
func (it *costAwareGroupIterator) Next() (int64, uint16, []byte, error) {
	for {
		if it.chunkIdx < 0 || it.chunkIdx >= len(it.indexChunks) {
			return 0, 0, nil, io.EOF
		}
		msgs := it.chunkMessages[it.chunkIdx]
		if it.msgIdx < 0 || it.msgIdx >= len(msgs) {
			it.advanceChunk()
			continue
		}

		// Materialize this chunk's compressed buffer on first touch.
		if it.isCompressed && it.loadedChunkIdx != it.chunkIdx {
			compressed := it.loadedData.Get(it.indexChunks[it.chunkIdx].ChunkOffset)
			if err := decompressInto(compressed, it.decompressBuf); err != nil {
				return 0, 0, nil, err
			}
			it.loadedChunkIdx = it.chunkIdx
		}

		msg := msgs[it.msgIdx]
		length := it.chunkMsgLens[it.chunkIdx][it.msgIdx]
		chunkOffset := it.indexChunks[it.chunkIdx].ChunkOffset

		var data []byte
		if it.isCompressed {
			end := msg.OffsetInChunk + length
			data = it.decompressBuf.Data[msg.OffsetInChunk:end]
		} else {
			data = it.loadedData.Get(chunkOffset + msg.OffsetInChunk)
		}

		it.advanceMessage()
		return msg.Timestamp, msg.TopicId, data, nil
	}
}

func (it *costAwareGroupIterator) advanceMessage() {
	if it.order == ReverseTimeOrder {
		it.msgIdx--
	} else {
		it.msgIdx++
	}
}

func (it *costAwareGroupIterator) advanceChunk() {
	if it.order == ReverseTimeOrder {
		it.chunkIdx--
		if it.chunkIdx >= 0 {
			it.msgIdx = len(it.chunkMessages[it.chunkIdx]) - 1
		}
	} else {
		it.chunkIdx++
		if it.chunkIdx < len(it.indexChunks) {
			it.msgIdx = 0
		}
	}
}

File: ts/src/iterators/cost_aware_group_iterator.ts
// Cost-aware per-group iterator: all bytes pre-fetched into LoadedBytes;
// per-chunk kept-message lists pre-computed at setup time. Mirrors
// go/cost_aware_group_iterator.go.

import type { Decompressor } from "../compression.js";
import type { LoadedBytes } from "../loaded_bytes.js";
import type { MessageRef } from "../sort_and_filter.js";
import type { IndexChunk } from "../types.js";

export class CostAwareGroupIterator {
  private readonly isCompressed: boolean;
  private readonly reverse: boolean;
  private readonly indexChunks: IndexChunk[];
  private readonly chunkMessages: MessageRef[][];
  private readonly chunkMsgLens: bigint[][];
  private readonly loadedData: LoadedBytes;
  private readonly decompress: Decompressor;

  // Decompressed buffer for the *current* compressed data chunk; undefined
  // when isCompressed=false or no chunk has been materialized yet.
  private currentDecompressed: Uint8Array | undefined;
  private loadedChunkIdx = -1;

  private chunkIdx: number;
  private msgIdx: number;

  constructor(args: {
    isCompressed: boolean;
    reverse: boolean;
    indexChunks: IndexChunk[];
    chunkMessages: MessageRef[][];
    chunkMsgLens: bigint[][];
    loadedData: LoadedBytes;
    decompress: Decompressor;
  }) {
    this.isCompressed = args.isCompressed;
    this.reverse = args.reverse;
    this.indexChunks = args.indexChunks;
    this.chunkMessages = args.chunkMessages;
    this.chunkMsgLens = args.chunkMsgLens;
    this.loadedData = args.loadedData;
    this.decompress = args.decompress;

    if (args.reverse) {
      this.chunkIdx = args.indexChunks.length - 1;
      this.msgIdx =
        this.chunkIdx >= 0
          ? args.chunkMessages[this.chunkIdx]!.length - 1
          : 0;
    } else {
      this.chunkIdx = 0;
      this.msgIdx = 0;
    }
  }

  /**
   * Next is synchronous (no I/O remaining). Returns undefined on EOF.
   * `data` aliases either loadedData or the per-chunk decompressed buffer.
   */
  next():
    | { timestamp: bigint; topicId: number; data: Uint8Array }
    | undefined {
    while (true) {
      if (this.chunkIdx < 0 || this.chunkIdx >= this.indexChunks.length) {
        return undefined;
      }
      const msgs = this.chunkMessages[this.chunkIdx]!;
      if (this.msgIdx < 0 || this.msgIdx >= msgs.length) {
        this.advanceChunk();
        continue;
      }

      // Materialize this chunk's decompressed bytes on first touch.
      if (this.isCompressed && this.loadedChunkIdx !== this.chunkIdx) {
        const compressed = this.loadedData.get(
          this.indexChunks[this.chunkIdx]!.chunkOffset,
        );
        if (compressed === undefined) {
          throw new Error(
            `cost-aware iterator: missing compressed chunk at offset ${this.indexChunks[this.chunkIdx]!.chunkOffset}`,
          );
        }
        this.currentDecompressed = this.decompress(compressed);
        this.loadedChunkIdx = this.chunkIdx;
      }

      const msg = msgs[this.msgIdx]!;
      const length = this.chunkMsgLens[this.chunkIdx]![this.msgIdx]!;
      const chunkOffset = this.indexChunks[this.chunkIdx]!.chunkOffset;

      let data: Uint8Array;
      if (this.isCompressed) {
        const start = Number(msg.offsetInChunk);
        const end = start + Number(length);
        data = this.currentDecompressed!.subarray(start, end);
      } else {
        const got = this.loadedData.get(chunkOffset + msg.offsetInChunk);
        if (got === undefined) {
          throw new Error(
            `cost-aware iterator: missing message bytes at offset ${chunkOffset + msg.offsetInChunk}`,
          );
        }
        data = got;
      }

      this.advanceMessage();
      return { timestamp: msg.timestamp, topicId: msg.topicId, data };
    }
  }

  private advanceMessage(): void {
    this.msgIdx += this.reverse ? -1 : 1;
  }

  private advanceChunk(): void {
    if (this.reverse) {
      this.chunkIdx -= 1;
      if (this.chunkIdx >= 0) {
        this.msgIdx = this.chunkMessages[this.chunkIdx]!.length - 1;
      }
    } else {
      this.chunkIdx += 1;
      if (this.chunkIdx < this.indexChunks.length) {
        this.msgIdx = 0;
      }
    }
  }
}
